File: route/view_w.py
from .tool.func import *

from .go_api_w_raw import api_w_raw
from .go_api_w_render import api_w_render
from .go_api_w_page_view import api_w_page_view

async def view_w(name = 'Test', do_type = ''):
    with get_db_connect() as conn:
        curs = conn.cursor()

        sub = 0
        history_color = 0
        menu = []

        user_doc = ''
        category_total = ''
        file_data = ''

        doc_type = ''
        redirect_to = None

        now_time = get_time()
        ip = ip_check()
            
        uppage = re.sub(r"/([^/]+)$", '', name)
        uppage = 0 if uppage == name else uppage

        curs.execute(db_change("select sub from rd where title = ? and not stop = 'O' order by date desc"), [name])
        topic = 1 if curs.fetchall() else 0

        curs.execute(db_change("select title from data where title like ?"), [name + '/%'])
        down = 1 if curs.fetchall() else 0

        if re.search(r'^category:', name):
            name_view = name
            doc_type = 'category'

            category_doc = ''
            category_sub = ''

            count_sub_category = 0
            count_category = 0

            curs.execute(db_change("select distinct link from back where title = ? and type = 'cat' order by link asc"), [name])
            category_sql = curs.fetchall()
            for data in category_sql:
                link_view = data[0]
                if get_main_skin_set(conn, flask.session, 'main_css_category_change_title', ip) != 'off':
                    curs.execute(db_change("select data from back where title = ? and link = ? and type = 'cat_view' limit 1"), [name, data[0]])
                    db_data = curs.fetchall()
                    if db_data and db_data[0][0] != '':
                        link_view = db_data[0][0]
                        
                link_blur = ''
                curs.execute(db_change("select data from back where title = ? and link = ? and type = 'cat_blur' limit 1"), [name, data[0]])
                db_data = curs.fetchall()
                if db_data:
                    link_blur = 'opennamu_category_blur'

                if data[0].startswith('category:'):
                    category_sub += '<li><a class="' + link_blur + '" href="/w/' + url_pas(data[0]) + '">' + html.escape(link_view) + '</a></li>'
                    count_sub_category += 1
                else:
                    category_doc += '' + \
                        '<li>' + \
                            '<a class="' + link_blur + '" href="/w/' + url_pas(data[0]) + '">' + html.escape(link_view) + '</a> ' + \
                            '<a class="opennamu_link_inter" href="/xref/' + url_pas(data[0]) + '">(' + get_lang(conn, 'backlink') + ')</a>' + \
                        '</li>' + \
                    ''
                    count_category += 1

            if category_sub != '':
                category_total += '' + \
                    '<h2 id="cate_under">' + get_lang(conn, 'under_category') + '</h2>' + \
                    '<ul>' + \
                        '<li>' + get_lang(conn, 'all') + ' : ' + str(count_sub_category) + '</li>' + \
                        category_sub + \
                    '</ul>' + \
                ''

            if category_doc != '':
                category_total += '' + \
                    '<h2 id="cate_normal">' + get_lang(conn, 'category_title') + '</h2>' + \
                    '<ul>' + \
                        '<li>' + get_lang(conn, 'all') + ' : ' + str(count_category) + '</li>' + \
                        category_doc + \
                    '</ul>' + \
                ''
        elif re.search(r"^user:([^/]*)", name):
            name_view = name
            doc_type = 'user'
            user_name = ''

            match = re.search(r"^user:([^/]*)", name)
            if match:
                user_name = html.escape(match.group(1))
            
            user_doc = ''

            # S admin or owner 특수 틀 추가
            if await acl_check(tool = 'all_admin_auth', ip = user_name) != 1:
                if await acl_check(tool = 'owner_auth', ip = user_name) != 1:
                    curs.execute(db_change('select data from other where name = "phrase_user_page_owner"'))
                    db_data = curs.fetchall()
                    if db_data and db_data[0][0] != '':
                        user_doc += db_data[0][0] + '<br>'
                    else:
                        curs.execute(db_change('select data from other where name = "phrase_user_page_admin"'))
                        db_data = curs.fetchall()
                        if db_data and db_data[0][0] != '':
                            user_doc += db_data[0][0] + '<br>'
                else:
                    curs.execute(db_change('select data from other where name = "phrase_user_page_admin"'))
                    db_data = curs.fetchall()
                    if db_data and db_data[0][0] != '':
                        user_doc += db_data[0][0] + '<br>'
            # E
            
            user_doc += '''
                <div id="opennamu_get_user_info">''' + html.escape(user_name) + '''</div>
                <hr class="main_hr">
            '''
            if name == 'user:' + user_name:
                menu += [['w/' + url_pas(name) + '/' + url_pas(now_time.split()[0]), get_lang(conn, 'today_doc')]]
        elif re.search(r"^file:", name):
            curs.execute(db_change('select id from history where title = ? order by date desc limit 1'), [name])
            db_data = curs.fetchall()
            rev = db_data[0][0] if db_data else '1' 

            name_view = name
            doc_type = 'file'

            mime_type = re.search(r'([^.]+)$', name)
            if mime_type:
                mime_type = mime_type.group(1)
            else:
                mime_type = 'jpg'

            file_name = re.sub(r'\.([^.]+)$', '', name)
            file_name = re.sub(r'^file:', '', file_name)

            file_all_name = sha224_replace(file_name) + '.' + mime_type
            file_path_name = os.path.join(load_image_url(conn), file_all_name)
            if os.path.exists(file_path_name):
                try:
                    img = Image.open(file_path_name)
                    width, height = img.size
                    file_res = str(width) + 'x' + str(height)
                except:
                    file_res = 'Vector'
                
                file_size = str(round(os.path.getsize(file_path_name) / 1000, 1))
                
                file_data = '''
                    <img src="/image/''' + url_pas(file_all_name) + '''.cache_v''' + rev + '''">
                    <h2>''' + get_lang(conn, 'data') + '''</h2>
                    <table>
                        <tr><td>''' + get_lang(conn, 'url') + '''</td><td><a href="/image/''' + url_pas(file_all_name) + '''">''' + get_lang(conn, 'link') + '''</a></td></tr>
                        <tr><td>''' + get_lang(conn, 'volume') + '''</td><td>''' + file_size + '''KB</td></tr>
                        <tr><td>''' + get_lang(conn, 'resolution') + '''</td><td>''' + file_res + '''</td></tr>
                    </table>
                    <h2>''' + get_lang(conn, 'content') + '''</h2>
                '''

                menu += [['delete_file/' + url_pas(name), get_lang(conn, 'file_delete')]]
            else:
                file_data = ''
        else:
            curs.execute(db_change("select link from back where title = ? and type = 'include' limit 1"), [name])
            doc_type = 'include' if curs.fetchall() else doc_type

            curs.execute(db_change("select title, data from back where link = ? and type = 'redirect' limit 1"), [name])
            db_data = curs.fetchall()
            if db_data:
                doc_type = 'redirect'

                curs.execute(db_change("select title from data where title = ?"), [db_data[0][0]])
                if curs.fetchall():
                    redirect_to = url_pas(db_data[0][0]) + db_data[0][1]

            name_view = name

        doc_data = await api_w_raw(name)
        if doc_data["response"] == "ok":
            render_data = await api_w_render(name, request_method = 'POST', request_data = {
                'name' : name,
                'data' : doc_data["data"]
            })
            end_data = render_data["data"] + '<script>document.addEventListener("DOMContentLoaded", function() {' + render_data["js_data"] + '});</script>'
        else:
            end_data = ''

        asyncio.create_task(api_w_page_view(name))

        curs.execute(db_change("select data from data where title = ?"), [name])
        data = curs.fetchall()

        description = ''
        if await acl_check(name, 'render') == 1:
            response_data = 401

            curs.execute(db_change('select data from other where name = "error_401"'))
            sql_d = curs.fetchall()
            if sql_d and sql_d[0][0] != '':
                end_data = '<h2>' + get_lang(conn, 'error') + '</h2><ul><li>' + sql_d[0][0] + '</li></ul>'
            else:
                end_data = '<h2>' + get_lang(conn, 'error') + '</h2><ul><li>' + get_lang(conn, 'authority_error') + '</li></ul>'
        elif not data:
            response_data = 404

            curs.execute(db_change('select data from other where name = "error_404"'))
            db_data = curs.fetchall()
            if db_data and db_data[0][0] != '':
                end_data = '<h2>' + get_lang(conn, 'error') + '</h2><ul><li>' + db_data[0][0] + '</li></ul>'
            else:
                end_data = '<h2>' + get_lang(conn, 'error') + '</h2><ul><li>' + get_lang(conn, 'decument_404_error') + '</li></ul>'

            curs.execute(db_change('select ip from history where title = ? limit 1'), [name])
            db_data = curs.fetchall()
            history_color = 1 if db_data else 0
        else:
            response_data = 200
            description = data[0][0].replace('\r', '').replace('\n', ' ')[0:200]

        curs.execute(db_change("select title from acl where title = ?"), [name])
        acl = 1 if curs.fetchall() else 0
        menu_acl = 1 if await acl_check(name, 'document_edit') == 1 else 0
        if response_data == 404:
            menu += [['edit/' + url_pas(name), get_lang(conn, 'create'), menu_acl]] 
        else:
            menu += [['edit/' + url_pas(name), get_lang(conn, 'edit'), menu_acl]]

        menu += [
            ['topic/' + url_pas(name), get_lang(conn, 'discussion'), topic], 
            ['history/' + url_pas(name), get_lang(conn, 'history'), history_color], 
            ['xref/' + url_pas(name), get_lang(conn, 'backlink')], 
            ['acl/' + url_pas(name), get_lang(conn, 'setting'), acl],
        ]

        if flask.session and 'lastest_document' in flask.session:
            if type(flask.session['lastest_document']) != type([]):
                flask.session['lastest_document'] = []
        else:
            flask.session['lastest_document'] = []

        if do_type == 'from':
            menu += [['w/' + url_pas(name), get_lang(conn, 'pass')]]
            
            last_page = ''
            for for_a in reversed(range(0, len(flask.session['lastest_document']))):
                last_page = flask.session['lastest_document'][for_a]

                curs.execute(db_change("select link from back where (title = ? or link = ?) and type = 'redirect' limit 1"), [last_page, last_page])
                if curs.fetchall():
                    break

            if last_page != name:
                redirect_text = '{0} ➤ {1}'

                curs.execute(db_change('select data from other where name = "redirect_text"'))
                db_data = curs.fetchall()
                if db_data and db_data[0][0] != '':
                    redirect_text = db_data[0][0]

                try:
                    redirect_text = redirect_text.format('<a href="/w_from/' + url_pas(last_page) + '">' + html.escape(last_page) + '</a>', '<b>' + html.escape(name) + '</b>')
                except:
                    redirect_text = '{0} ➤ {1}'
                    redirect_text = redirect_text.format('<a href="/w_from/' + url_pas(last_page) + '">' + html.escape(last_page) + '</a>', '<b>' + html.escape(name) + '</b>')

                end_data = '''
                    <div class="opennamu_redirect" id="redirect">
                        ''' + redirect_text + '''
                    </div>
                    <hr class="main_hr">
                ''' + end_data
                
        if len(flask.session['lastest_document']) >= 10:
            flask.session['lastest_document'] = flask.session['lastest_document'][-9:] + [name]
        else:
            flask.session['lastest_document'] += [name]
        
        flask.session['lastest_document'] = list(reversed(dict.fromkeys(reversed(flask.session['lastest_document']))))

        if redirect_to and do_type != 'from':
            return redirect(conn, '/w_from/' + redirect_to)

        view_history_on = get_main_skin_set(conn, flask.session, 'main_css_view_history', ip)
        if view_history_on == 'on':
            end_data = '' + \
                '<div class="opennamu_trace">' + \
                    '<a class="opennamu_trace_button" href="javascript:opennamu_do_trace_spread();"> (+)</a>' + \
                    get_lang(conn, 'trace') + ' : ' + \
                    ' ← '.join(
                        [
                            '<a href="/w/' + url_pas(for_a) + '">' + html.escape(for_a) + '</a>'
                            for for_a in reversed(flask.session['lastest_document'])
                        ]
                    ) + \
                '</div>' + \
                '<hr class="main_hr">' + \
            '' + end_data

        if uppage != 0:
            menu += [['w/' + url_pas(uppage), get_lang(conn, 'upper')]]

        if down:
            menu += [['down/' + url_pas(name), get_lang(conn, 'sub')]]

        curs.execute(db_change("select set_data from data_set where doc_name = ? and set_name = 'last_edit'"), [name])
        r_date = curs.fetchall()
        r_date = r_date[0][0] if r_date else 0

        curs.execute(db_change("select data from other where name = 'not_use_view_count'"))
        view_count_set = curs.fetchall()
        if view_count_set and view_count_set[0][0] != "":
            view_count = 0
        else:
            curs.execute(db_change("select set_data from data_set where doc_name = ? and set_name = 'view_count' and doc_rev = ''"), [name])
            view_count = curs.fetchall()
            view_count = view_count[0][0] if view_count else 0

        div = file_data + user_doc + end_data + category_total
        
        if doc_type == '':
            curs.execute(db_change('select data from other where name = "outdated_doc_warning_date"'))
            db_data = curs.fetchall()
            if db_data and db_data[0][0] != '' and r_date != 0:
                time_1 = datetime.datetime.strptime(r_date, '%Y-%m-%d %H:%M:%S') + datetime.timedelta(days = int(number_check(db_data[0][0])))
                time_2 = datetime.datetime.strptime(now_time, '%Y-%m-%d %H:%M:%S')
                if time_2 > time_1:
                    curs.execute(db_change('select data from other where name = "outdated_doc_warning"'))
                    db_data = curs.fetchall()
                    div = (db_data[0][0] if db_data and db_data[0][0] != '' else get_lang(conn, 'old_page_warning')) + '<hr class="main_hr">' + div

        curs.execute(db_change("select data from other where name = 'body'"))
        body = curs.fetchall()
        div = (body[0][0] + div) if body else div

        curs.execute(db_change("select data from other where name = 'bottom_body'"))
        body = curs.fetchall()
        div += body[0][0] if body else ''

        curs.execute(db_change("select set_data from data_set where doc_name = ? and set_name = 'document_top'"), [name])
        body = curs.fetchall()
        div = (body[0][0] + div) if body else div

        if ip_or_user(ip) == 0:
            curs.execute(db_change("select data from user_set where id = ? and data = ?"), [ip, name])
            watch_list = 2 if curs.fetchall() else 1
            menu += [['star_doc_from/' + url_pas(name), ('☆' if watch_list == 1 else '★'), watch_list - 1]]
        else:
            watch_list = 0

        menu += [['doc_watch_list/1/' + url_pas(name), get_lang(conn, 'watchlist')]]

        return easy_minify(conn, flask.render_template(skin_check(conn),
            imp = [name_view, await wiki_set(), await wiki_custom(conn), wiki_css([sub, r_date, watch_list, description, view_count])],
            data = div,
            menu = menu
        )), response_data

File: route/view_w.go
package route

import (
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"time"

	"opennamu/route/tool"

	"github.com/gin-gonic/gin"
)

var view_w_image_regex = regexp.MustCompile(`(?is)<img\b[^>]*\bsrc\s*=\s*"([^"]+)"`)

func view_w_get_first_image(data string) string {
	match := view_w_image_regex.FindStringSubmatch(data)
	if len(match) < 2 {
		return ""
	}

	return tool.HTML_unescape(match[1])
}

func View_w(c *gin.Context, config tool.Config, doc_name string, view_type string) (string, int) {
	db := tool.DB_connect()
	defer tool.DB_close(db)

	do_type := view_type

	Api_w_page_view_post(config, doc_name)

	var render_data string
	status := http.StatusOK
	raw_data := ""
	length_doc := ""
	description := ""
	og_image := ""

	raw_data_api := Api_w_raw(config, doc_name, "", "")
	raw_response, _ := raw_data_api["response"].(string)
	if raw_response == "require auth" {
		error_data := tool.Get_setting_value(db, "error_401", "", "")
		if error_data != "" {
			render_data = "<h2>" + tool.Get_language(db, "error", true) + "</h2><ul><li>" + error_data + "</li></ul>"
		} else {
			render_data = "<h2>" + tool.Get_language(db, "error", true) + "</h2><ul><li>" + tool.Get_language(db, "authority_error", true) + "</li></ul>"
		}
		status = http.StatusUnauthorized
	} else if raw_response != "ok" {
		error_data := tool.Get_setting_value(db, "error_404", "", "")
		if error_data != "" {
			render_data = "<h2>" + tool.Get_language(db, "error", true) + "</h2><ul><li>" + error_data + "</li></ul>"
		} else {
			render_data = "<h2>" + tool.Get_language(db, "error", true) + "</h2><ul><li>" + tool.Get_language(db, "document_404_error", true) + "</li></ul>"
		}
		status = http.StatusNotFound
	} else {
		if do_type != "from" {
			redirect_target, redirect_anchor := view_w_redirect_target(db, doc_name)
			if redirect_target != "" {
				view_w_add_recent_document(config, doc_name)
				return tool.Get_redirect("/w_from/" + tool.Url_parser(redirect_target) + redirect_anchor), http.StatusOK
			}
		}

		raw_data, _ = raw_data_api["data"].(string)
		render_data_api := Api_w_render(config, doc_name, raw_data, "normal", "")
		render_data, _ = render_data_api["data"].(string)
		og_image = view_w_get_first_image(render_data)
		if strings.HasPrefix(og_image, "/") {
			domain := tool.Get_domain(db, true)
			if domain == "http://" || domain == "https://" {
				protocol := "http"
				if c.Request.TLS != nil || strings.EqualFold(c.GetHeader("X-Forwarded-Proto"), "https") {
					protocol = "https"
				}
				domain = protocol + "://" + c.Request.Host
			}
			if domain != "" && domain != "http://" && domain != "https://" {
				og_image = strings.TrimRight(domain, "/") + og_image
			}
		} else if !strings.HasPrefix(og_image, "http://") && !strings.HasPrefix(og_image, "https://") {
			og_image = ""
		}
		length_doc = strconv.Itoa(tool.Get_len(raw_data))
		description = tool.Get_slice(strings.ReplaceAll(strings.ReplaceAll(raw_data, "\r", ""), "\n", " "), 0, 200)
	}

	recent_documents := view_w_recent_documents(config)
	if do_type == "from" {
		render_data = view_w_redirect_trace(db, doc_name, recent_documents) + render_data
	}
	recent_documents = view_w_add_recent_document(config, doc_name)
	render_data = view_w_trace(db, config, recent_documents) + render_data
	document_type := ""
	if strings.HasPrefix(doc_name, "user:") {
		document_type = "special"
		render_data = view_w_user_data(db, doc_name) + render_data
	} else if strings.HasPrefix(doc_name, "category:") {
		document_type = "special"
		render_data += view_w_category_data(db, config, doc_name)
	} else if strings.HasPrefix(doc_name, "file:") {
		document_type = "special"
		render_data = view_w_file_data(db, doc_name) + render_data
	} else {
		if _, exists := tool.Get_back_link(db, doc_name, "include"); exists {
			document_type = "include"
		}
		if _, exists := tool.Get_back_title(db, doc_name, "redirect"); exists {
			document_type = "redirect"
		}
	}

	last_edit := tool.Get_document_setting_value_exact(db, doc_name, "last_edit", "")

	if document_type == "" && last_edit != "" {
		warning_days := tool.Get_setting_value(db, "outdated_doc_warning_date", "", "")
		warning_days_int := tool.Str_to_int(warning_days)
		if warning_days_int > 0 {
			last_edit_time, parse_error := time.Parse("2006-01-02 15:04:05", last_edit)
			if parse_error == nil && time.Now().After(last_edit_time.AddDate(0, 0, warning_days_int)) {
				warning_text := tool.Get_setting_value(db, "outdated_doc_warning", "", "")
				if warning_text == "" {
					warning_text = tool.Get_language(db, "old_page_warning", true)
				}
				render_data = warning_text + `<hr class="main_hr">` + render_data
			}
		}
	}

	body := tool.Get_setting_value(db, "body", "", "")
	render_data = body + render_data

	bottom_body := tool.Get_setting_value(db, "bottom_body", "", "")
	render_data += bottom_body

	document_top := tool.Get_document_setting_value_exact(db, doc_name, "document_top", "")
	render_data = document_top + render_data

	topic := 0
	if tool.Get_rd_active_title(db, doc_name) {
		topic = 1
	}
	history_color := 0
	if status == http.StatusNotFound {
		if tool.Get_history_exists(db, doc_name) {
			history_color = 1
		}
	}
	acl_color := 0
	if tool.Get_acl_exists(db, doc_name) {
		acl_color = 1
	}
	menu_acl := 0
	if tool.Check_acl(db, doc_name, "", "document_edit", config.IP) {
		menu_acl = 1
	}

	watch_list := 0
	menu := [][]any{
		{"edit/" + tool.Url_parser(doc_name), tool.Get_language(db, "edit", true), menu_acl},
		{"topic/" + tool.Url_parser(doc_name), tool.Get_language(db, "discussion", true), topic},
		{"history/" + tool.Url_parser(doc_name), tool.Get_language(db, "history", true), history_color},
		{"xref/" + tool.Url_parser(doc_name), tool.Get_language(db, "backlink", true)},
		{"acl/" + tool.Url_parser(doc_name), tool.Get_language(db, "setting", true), acl_color},
	}
	if status == http.StatusNotFound {
		menu[0][1] = tool.Get_language(db, "create", true)
	}
	if do_type == "from" {
		menu = append(menu, []any{"w/" + tool.Url_parser(doc_name), tool.Get_language(db, "pass", true)})
	}
	if strings.HasPrefix(doc_name, "user:") && !strings.Contains(strings.TrimPrefix(doc_name, "user:"), "/") {
		menu = append(menu, []any{"w/" + tool.Url_parser(doc_name+"/"+tool.Get_date()), tool.Get_language(db, "today_doc", true)})
	}
	if strings.HasPrefix(doc_name, "file:") {
		menu = append(menu, []any{"delete_file/" + tool.Url_parser(doc_name), tool.Get_language(db, "file_delete", true)})
	}
	if slash_index := strings.LastIndex(doc_name, "/"); slash_index > 0 {
		menu = append(menu, []any{"w/" + tool.Url_parser(doc_name[:slash_index]), tool.Get_language(db, "upper", true)})
	}
	if view_w_child_exists(db, doc_name) {
		menu = append(menu, []any{"down/" + tool.Url_parser(doc_name), tool.Get_language(db, "sub", true)})
	}
	if !tool.IP_or_user(config.IP) {
		if tool.Get_user_set_data_match(db, config.IP, "star_doc", doc_name) {
			watch_list = 2
		} else {
			watch_list = 1
		}
		star_text := "☆"
		if watch_list == 2 {
			star_text = "★"
		}
		menu = append(menu, []any{"star_doc_from/" + tool.Url_parser(doc_name), star_text + " " + tool.Get_language(db, "star_doc", true), watch_list - 1})
	}
	if tool.Check_permission(db, "doc_watch_list_view", config.IP) {
		menu = append(menu, []any{"doc_watch_list/1/" + tool.Url_parser(doc_name), tool.Get_language(db, "watch_user_list", true)})
	}

	enable_comment := tool.Get_setting(db, "enable_comment", "")
	if status == http.StatusOK && len(enable_comment) > 0 && enable_comment[0][0] != "" {
		comment_api := Api_w_comment_ui(config, doc_name)
		if comment_data, ok := comment_api["data"].(string); ok && comment_data != "" {
			render_data += `<div class="opennamu_clearfix"></div>` + comment_data
		}
	}

	view_count := 0
	view_count_api := Api_w_page_view(config, doc_name)
	if view_count_data, ok := view_count_api["data"].(int); ok {
		view_count = view_count_data
	}

	out := tool.Get_template(
		db,
		config,
		doc_name,
		render_data,
		[]any{"", last_edit, watch_list, description, view_count},
		menu,
		map[string]string{
			"path":       c.Request.URL.Path,
			"length_doc": length_doc,
			"og_image":   og_image,
		},
	)

	return out, status
}
